README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![R build
status](https://github.com/obrl-soil/h3jsr/workflows/R-CMD-check/badge.svg)](https://github.com/obrl-soil/h3jsr/actions)
[![R build
status](https://github.com/obrl-soil/h3jsr/workflows/test-coverage/badge.svg)](https://github.com/obrl-soil/h3jsr/actions)
[![R build
status](https://github.com/obrl-soil/h3jsr/workflows/pkgdown/badge.svg)](https://github.com/obrl-soil/h3jsr/actions)
[![Coverage
status](https://codecov.io/gh/obrl-soil/h3jsr/branch/master/graph/badge.svg)](https://codecov.io/github/obrl-soil/h3jsr?branch=master)
[![CRAN](https://www.r-pkg.org/badges/version/h3jsr)](https://cran.r-project.org/package=h3jsr)
[![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/h3jsr)](https://www.r-pkg.org/pkg/h3jsr)

# h3jsr

h3jsr provides access to [Uber’s H3 library](https://github.com/uber/h3)
via its [javascript transpile](https://github.com/uber/h3-js), using the
magical power of [`V8`](https://github.com/jeroen/v8).

H3 is a hexagonal hierarchical geospatial indexing system. Details about
its structure and use cases [can be found
here](https://h3geo.org/docs/).

## Installation

Install from CRAN with

``` r
install.packages('h3jsr')
```

Install the development version from GitHub with

``` r
remotes::install_github("obrl-soil/h3jsr")
```

> :bulb: Version 1.3.0 contains an API revision, so some functions
> have new names. See the NEWS.

## Example

``` r
library(h3jsr)
library(sf)
#> Linking to GEOS 3.9.3, GDAL 3.5.2, PROJ 8.2.1; sf_use_s2() is TRUE

# where is the Brisbane Town Hall at resolution 15?
bth <- st_sfc(st_point(c(153.023503, -27.468920)), crs = 4326)
point_to_cell(bth, res = 15)
#> [1] "8fbe8d12acad2f3"

# where is it at several resolutions?
point_to_cell(bth, res = seq(10, 15), simple = FALSE)
#>   h3_resolution_10 h3_resolution_11 h3_resolution_12 h3_resolution_13
#> 1  8abe8d12acaffff  8bbe8d12acadfff  8cbe8d12acad3ff  8dbe8d12acad2ff
#>   h3_resolution_14 h3_resolution_15
#> 1  8ebe8d12acad2f7  8fbe8d12acad2f3

# Where is the center of the hexagon over the Brisbane Town 
# Hall at resolution 10?
brisbane_10 <- cell_to_point(h3_address = '8abe8d12acaffff')
brisbane_10
#> Geometry set for 1 feature 
#> Geometry type: POINT
#> Dimension:     XY
#> Bounding box:  xmin: 153.0239 ymin: -27.46853 xmax: 153.0239 ymax: -27.46853
#> Geodetic CRS:  WGS 84
#> POINT (153.0239 -27.46853)

# Is that a valid H3 address?
is_valid(h3_address = '8abe8d12acaffff')
#> [1] TRUE

# is it a pentagon?
is_pentagon(h3_address = '8abe8d12acaffff')
#> [1] FALSE

# is it Class III?
is_rc3(h3_address = '8abe8d12acaffff')
#> [1] FALSE

# What is Brisbane Town Hall's base cell number?
get_base_cell(h3_address = '8abe8d12acaffff')
#> [1] 95

# What is the hexagon over the Brisbane Town Hall at resolution 10?
brisbane_hex_10 <- cell_to_polygon(input = '8abe8d12acaffff', simple = FALSE)

# if you're feeling fancy,
# point_to_cell(bth, res = seq(10,15)) %>%
#   unlist() %>%
#   h3_to_polygon(., simple = FALSE) %>%
#   mapview::mapview()
  
```

Props to Joel Gombin, who’s package
[`concaveman`](https://github.com/joelgombin/concaveman) provided me
with the implementation inspo.

------------------------------------------------------------------------


NEWS.md
# Version 1.3.1 [CRAN] 

  * Updated to h3-js 4.1.0
  * New functions `cell_to_childpos()`, `childpos_to_cell()` and `cell_to_children_size()`. See this [Observable notebook](https://observablehq.com/d/697d0ae050effa18) for use cases.

# Version 1.3.0

  * Updated to h3-js 4.0.1. Note that v4+ has extensive changes to the underlying API. Some functions in this package have been renamed for consistency:
    * `point_to_h3()` --> `point_to_cell()`
    * `h3_to_point()` --> `cell_to_point()`
    * `h3_to_polygon()` --> `cell_to_polygon()`
    * `h3_to_line()` --> `cell_to_line()`
    * `get_kring()` --> `get_disk()`
    * `get_kring_list()` --> `get_disk_list()`
    * `polyfill()` --> `polygon_to_cells()`
    * `set_to_multipolygon()` --> `cells_to_multipolygon()`
    * `res_count()` --> `num_cells()`
  * New functions `is_valid_vertex()`, `get_cell_vertex()`, `get_cell_vertexes()`, and `vertex_to_point()` are available for interacting with H3 in vertex mode.
  * New functions `cell_to_splitlong()` and `splitlong_to_cell()` are available for converting between 64-bit string addresses and 32-bit integer pairs.
  * New functions `degs_to_rads()` and `rads_to_degs()` available for unit conversion.

# Version 1.2.3 [CRAN] 

  * Updated `V8` version requirement to ensure consistent build across platforms

# Version 1.2.2 [CRAN]

  * Updated to h3-js 3.7.2

# Version 1.2.1 

  * Updated to h3-js 3.7.1

# Version 1.2.0 

  * Updated to h3-js 3.7.0
  * Added functions `cell_area`, `edge_length`, `get_dist`, and `get_res0`.
  * Revised internal function `prep_for_pt2h3`

# Version 1.1.3 

  * Updated to h3-js 3.6.4

# Version 1.1.2 

  * Updated to h3-js 3.6.3 - polyfill algo was rolled back due to known issues
  * Move to Github Actions for CI workflow
  * Remove unneeded lwgeom dependency
  * Bugfix for accessing `h3_info_table` when library not loaded explicitly

# Version 1.1.1 

  * Updated to h3-js 3.6.2 - expect improvements to `polyfill()` behaviour

# Version 1.1

  * Updated to h3-js 3.6.1
  * Added `get_centerchild()`, which will return the central child of a hexagon at a given resolution.
  * Added `get_pentagons()`, which will return the twelve pentagon indexes at a given resolution.
  
# Version 1.0

  * Updated to h3-js 3.5.0, now using official browser bundle
  * Added `get_faces()`, which returns the icosahedron face(s) for a given H3 address.

# Version 0.9

  * Updated to h3-js 3.4.2  
  * Replaced custom function `min_path()` with official version `grid_path()`, which wraps `h3Line` to find a path between 
    two addresses.
  * Added custom function `h3_to_line()`, which converts a vector of H3 addresses to `sfc_LINESTRING`.

# Version 0.8

  * Updated to h3-js 3.3.0 and babel-polyfill 6.26.0

# Version 0.7

  * Removed `nearest_neighbour()`, its fundamentally flawed. 
  * Simplified output of `get_local_ij()` to a matrix when simple = TRUE
  * Improved outputs from `h3_to_point()`
  * Efficiency improvements to `h3_to_point()` and `polyfill()`
  * `h3_to_point()`, `h3_to_polygon()`, and `polyfill()` now take a wider range of input objects.

# Version 0.6

  * Update core library to v 3.2.0
  * Add `get_local_ij()` and `get_local_h3()`, wrapping the experimental local coordinate system functions `h3.experimentalH3ToLocalIj` and `h3.experimentalLocalIjToH3` respectively.

# Version 0.5.1
  
  * Added custom function `min_path()` to find a minimum-steps path between two H3 addresses of the same resolution.
  * Added custom function `nearest_neighbour()` to find the nearest neighbour for a given set of points. Use with caution, read the notes.

# Version 0.5.0
 
  * change to v8 session handling (speed enhancement)

# Version 0.4.5

  * update core library to v 3.1.1

# Version 0.4.4
 
  * bugfix for `polyfill()`. Where input `sf` object only has a geometry column, a column called 'ID_H3' containing a sequential ID field is now added before sending the geometry to the V8 session.

# Version 0.4.3

  * Update h3-js to release 3.1.0 - https://github.com/uber/h3-js/releases/tag/v3.1.0
  * Add `grid_distance()`, which returns the number of 'steps' required to get from one H3 address to another (within the same resolution).

# Version 0.4.2

  * Update h3-js to release 3.0.2 - https://github.com/uber/h3-js/releases/tag/v3.0.2

# Version 0.4.1

  * Warnings converted to messages
  * Multipolygon support for `h3_polyfill()`, plus a message when output resolution is way smaller than input extent.

# Version 0.4.0

  * Naming things is hard, but its vignette time so, y'know, now or never. Function names are shorter now and make more sense.
  * Added another info utility `res_cendist()` for getting the average distance between the center of each region represented by a H3 address at a given resolution.
  * Vignette covering core functionality added.

# Version 0.3.1
  
  * Reworked `to_point()` to return an `sfc_POINT` object, or an `sf` point data frame where `simple = FALSE`.
  * Improved outputs of `to_polygon()` to have a similar structure to `h3_to_geo()`. 
  
# Version 0.3.0

  * Reworked `point_to_h3()` to take in an `sf` point object and return the same when `simple = FALSE`, resulting in a new dependency on `tidyr`. On the upside, one can now request addresses for multiple resolutions at multiple points, if one wishes to test the limits of one's system.

# Version 0.2.3

  * Added public informational utilities `res_area()`, `res_edgelen()`, and `res_count()`
  * Added data table of h3 address info for fast retrieval of information using the above functions
  * Fixed resolution validation bug, level 0 is now allowed

# Version 0.2.2

  * All unidirectional algorithms added: `are_neighbours()`, `get_udedge()`, `is_valid_edge()`, `get_udorigin()`, `get_uddest()`, `get_udends()`, `get_udedges()`, and `udedge_to_line()`.

# Version 0.2.1
  
  * `polyfill()` and `set_to_multipolygon()` added; new dependencies on `sf`and `geojsonsf` have resulted.

R/info_utilities.R
#' Get average cell area
#'
#' This function returns the average area of an H3 cell at a given
#' resolution.
#' @inheritParams get_parent
#' @param units Areal unit to report in. Options are square meters or square
#'   kilometers.
#' @param fast Logical; whether to retrieve values from a locally stored table or
#'   reclaculate from source.
#' @return Numeric; average H3 cell area.
#' @examples
#' # Return average H3 cell area at each resolution in square meters
#' res_area(res = seq(0, 15), units = 'm2')
#'
#' @import V8
#' @importFrom utils data
#' @export
#'
res_area <- function(res = NULL, units = c('m2', 'km2'), fast = TRUE) {

  if(!any(res %in% seq(0, 15))) {
    stop('Please provide a valid H3 resolution. Allowable values are 0-15 inclusive.')
  }

  units <-  match.arg(units) # NB rads unsupported at 2022-09, see h3-js #134

  if(fast == TRUE) {
    utils::data('h3_info_table', envir = environment(), package = 'h3jsr')
    h3_info_table <- h3_info_table[h3_info_table$h3_resolution %in% res,
                                   switch(units, 'm2'  = 'avg_area_sqm',
                                                 'km2' = 'avg_area_sqkm')]
    return(h3_info_table)
  } else {
    sesh$assign('evalThis', data.frame(res))
    sesh$assign('unit', units)
    # sesh$eval('console.log(unit);')
    # sesh$eval('console.log(JSON.stringify(h3.getHexagonAreaAvg(evalThis[0].res, unit)));')
    sesh$eval('for (var i = 0; i < evalThis.length; i++) {
                 evalThis[i].area = h3.getHexagonAreaAvg(evalThis[i].res, unit);
              };')
    sesh$get('evalThis')
  }
}

#' Get exact cell area
#'
#' This function calculates the exact area of an H3 cell.
#' @param h3_address Character; 15-character index generated by H3.
#' @param simple Logical; whether to return a vector of outputs or a data frame
#'   containing both inputs and outputs.
#' @param units Length unit to report in. Options are square meters, square
#'   kilometers, or steradians.
#' @return By default, a numeric vector of length(h3_address).
#' @examples
#' cell_area(h3_address = '8abe8d12acaffff', 'm2')
#' @import V8
#' @export
#'
cell_area <- function(h3_address = NULL,
                      units = c('m2', 'km2', 'rads2'), simple = TRUE) {

  units <- match.arg(units, c('m2', 'km2', 'rads2'))

  # edge addresses don't fail, they just return nonsense so:
  stopifnot(all(is_valid_edge(h3_address)) == FALSE)

  eval_this <- data.frame(h3_address, stringsAsFactors = FALSE)

  sesh$assign('evalThis', eval_this)
  sesh$assign('units', units)

  # for debug:
  # sesh$eval('console.log(JSON.stringify(evalThis[0]))')
  # sesh$eval('console.log(JSON.stringify(h3.cellArea(evalThis[0].h3_address)));')
  sesh$eval('for (var i = 0; i < evalThis.length; i++) {
            evalThis[i].area = h3.cellArea(evalThis[i].h3_address, units);
            };')

  # retrieve the result
  if(simple == TRUE) {
    sesh$get('evalThis')$area
  } else {
    out <- sesh$get('evalThis')
    names(out) <- switch(units,
                         'm2'    = c('h3_address', 'area_m2'),
                         'km2'   = c('h3_address', 'area_km2'),
                         'rads2'  = c('h3_address', 'area_rads2'))
    out
  }

}

#' Get average cell edge length
#'
#' This function returns the average edge length of an H3 cell edge at a given
#' resolution.
#' @inheritParams get_parent
#' @param units Length unit to report in. Options are meters or kilometers.
#' @param fast Logical; whether to retrieve values from a locally stored table
#'   or recalculate from source.
#' @return Numeric; H3 cell edge length
#' @note This value is also the hexagon circumradius.
#' @examples
#' # Return average H3 cell edge length at each resolution in kilometers
#' res_length(res = seq(0, 15), units = 'km')
#'
#' @import V8
#' @importFrom utils data
#' @export
#'
res_length <- function(res = NULL, units = c('m', 'km'), fast = TRUE) {

  if(!any(res %in% seq(0, 15))) {
    stop('Please provide a valid H3 resolution. Allowable values are 0-15 inclusive.')
  }

  units <-  match.arg(units) # NB rads unsupported at 2022-09, see h3-js #134

  if(fast == TRUE) {
    utils::data('h3_info_table', envir = environment(), package = 'h3jsr')
    h3_info_table <- h3_info_table[h3_info_table$h3_resolution %in% res,
                                   switch(units, 'm'    = 'avg_edge_m',
                                                 'km'   = 'avg_edge_km')]
    return(h3_info_table)
  } else {
  sesh$assign('evalThis', data.frame(res))
  sesh$assign('unit', match.arg(units))
  # sesh$eval('console.log(unit);')
  # sesh$eval('console.log(JSON.stringify(h3.getHexagonEdgeLengthAvg(evalThis[0].res, unit));')
  sesh$eval('for (var i = 0; i < evalThis.length; i++) {
               evalThis[i].edgelen = h3.getHexagonEdgeLengthAvg(evalThis[i].res, unit);
            };')
  sesh$get('evalThis')
  }

}

#' Get exact cell edge length
#'
#' This function calculates the exact length of an H3 cell edge.
#' @param h3_edge Character; address of unidirectional edge.
#' @param simple Logical; whether to return a vector of outputs or a data frame
#'   containing both inputs and outputs.
#' @param units Length unit to report in. Options are meters, kilometers, or
#'   radians.
#' @return By default, a numeric vector of length(h3_address).
#' @examples
#' edge_length(h3_edge = '166be8d12fffffff', 'm')
#' @import V8
#' @export
#'
edge_length <- function(h3_edge = NULL, units = c('m', 'km', 'rads'),
                        simple = TRUE) {

  units <- match.arg(units, c('m', 'km', 'rads'))

  stopifnot(all(is_valid_edge(h3_edge)) == TRUE)

  eval_this <- data.frame(h3_edge, stringsAsFactors = FALSE)

  sesh$assign('evalThis', eval_this)
  sesh$assign('units', units)

  # for debug:
  # sesh$eval('console.log(JSON.stringify(evalThis[0]))')
  # sesh$eval('console.log(JSON.stringify(h3.edgeLength(evalThis[i].h3_edge, units)));')
  sesh$eval('for (var i = 0; i < evalThis.length; i++) {
            evalThis[i].len = h3.edgeLength(evalThis[i].h3_edge, units);
            };')

  # retrieve the result
  if(simple == TRUE) {
    sesh$get('evalThis')$len
  } else {
    out <- sesh$get('evalThis')
    names(out) <- switch(units,
                         'm'    = c('h3_address', 'length_m'),
                         'km'   = c('h3_address', 'length_km'),
                         'rads' = c('h3_address', 'length_rads'))
    out
    }

}

#' Get average distance between H3 cell centers
#'
#' This function returns the average distance between the center of H3 cells
#'  at a given resolution.
#' @inheritParams get_parent
#' @param units Length unit to report in, either meters or kilometers.
#' @param fast Logical; whether to retrieve values from a locally stored table or
#'   recalculate from source.
#' @return Numeric; H3 cell center separation distance.
#' @note This isn't in the core library but may be useful.
#' @examples
#' # Return average H3 cell separation distance at each resolution in kilometers
#' res_cendist(res = seq(0, 15), units = 'km')
#'
#' @import V8
#' @importFrom utils data
#' @export
#'
res_cendist <- function(res = NULL, units = c('m', 'km'), fast = TRUE) {

  if(!any(res %in% seq(0, 15))) {
    stop('Please provide a valid H3 resolution. Allowable values are 0-15 inclusive.')
  }

  units <-  match.arg(units) # rads unsupported at present

  if(fast == TRUE) {
    utils::data('h3_info_table', envir = environment(), package = 'h3jsr')
    h3_info_table <- h3_info_table[h3_info_table$h3_resolution %in% res,
                                   switch(units, 'm'    = 'avg_cendist_m',
                                                 'km'   = 'avg_cendist_km')]
    return(h3_info_table)
  } else {
    crad <- h3jsr::res_length(res = res, units = units, fast = FALSE)
    crad$cendist <- cos(30 * pi / 180) * crad$edgelen * 2
    crad[, c('res', 'cendist')]
  }

}

#' Get total H3 cells
#'
#' This function returns total number of H3 cells at a given resolution.
#' @inheritParams get_parent
#' @param fast Logical; whether to retrieve values from a locally stored table or
#'   recalculate from source.
#' @return Numeric; H3 cell count.
#' @note Above resolution 8 the exact count cannot be represented in a
#'   JavaScript 32-bit number, so consumers should use caution when applying
#'   further operations to the output.
#' @examples
#' # Return cell count for resolution 8
#' num_cells(res = 8)
#'
#' @import V8
#' @importFrom utils data
#' @export
#'
num_cells <- function(res = NULL, fast = TRUE) {

  if(!any(res %in% seq(0, 15))) {
    stop('Please provide a valid H3 resolution. Allowable values are 0-15 inclusive.')
  }

  if(fast == TRUE) {
    utils::data('h3_info_table', envir = environment(), package = 'h3jsr')
    h3_info_table <- h3_info_table[h3_info_table$h3_resolution %in% res,
                                   'total_unique_indexes']
    return(h3_info_table)
  } else {
    sesh$assign('evalThis', data.frame(res))
    # sesh$eval('console.log(unit);')
    # sesh$eval('console.log(JSON.stringify(h3.getNumCells(evalThis[0].res)));')
    sesh$eval('for (var i = 0; i < evalThis.length; i++) {
            evalThis[i].total_unique_indexes = h3.getNumCells(evalThis[i].res);
            };')
    sesh$get('evalThis')
  }
}

#' Great circle distance
#'
#' Get the great circle distance between WGS84 lat/long points
#' @param pt1 `sf` object with point geometry, `sfc_POINT` object, `sfg`
#'   point, data frame or matrix.
#' @param pt2 `sf` object with point geometry, `sfc_POINT` object, `sfg`
#'   point, data frame or matrix.
#' @param units whether to return the great circle distance in meters,
#'   kilometers, or radians.
#' @param simple whether to return a numeric vector of distances or a
#'   `data.frame` containing start and end coordinates as well as distance.
#' @return Numeric vector of point to point distances, or data frame of origin
#'   and destination coordinates accompanied by their distances.
#' @note This functionality also exists in R packages \code{sp}, \code{sf},
#' \code{geosphere} and \code{fields}. H3's version appears to return slightly
#' shorter distances than most other implementations, but is included here
#' for completeness.
#' @examples
#' # distance between Brisbane and Melbourne
#' bne <- c(153.028, -27.468)
#' mlb <- c(144.963, -37.814)
#' get_gcdist(bne, mlb, 'km')
#' @export
#'
get_gcdist <- function(pt1 = NULL, pt2 = NULL,
                       units = c('m', 'km', 'rads'),
                       simple = TRUE) {

  units <- match.arg(units, c('m', 'km', 'rads'))

  pt1 <- prep_for_pt2cell(pt1)
  pt2 <- prep_for_pt2cell(pt2)

  stopifnot(nrow(pt1) == nrow(pt2)) # TODO

  sesh$assign('pt1', pt1, digits = NA)
  sesh$assign('pt2', pt2, digits = NA)
  sesh$assign('units', units, digits = NA)

  # sesh$eval('console.log(pts1[0]);')
  # sesh$eval('console.log(JSON.stringify(h3.greatCircleDistance(pts1[0], pts2[0], units)));')
  sesh$eval('var gcd = [];
            for (var i = 0; i < pt1.length; i++) {
              gcd[i] = h3.greatCircleDistance(pt1[i], pt2[i], units);
            };')

  if(simple == TRUE) {
    sesh$get('gcd')
  } else {
    out <- as.data.frame(cbind(pt1, pt2, sesh$get('gcd')))
    names(out) <- c('start_X', 'start_Y', 'end_X', 'end_Y', 'gc_dist')
    out
  }
}

# ## get all info in a table for fast access
# h3_res_areas <-
#   purrr::reduce(list(res_area(seq(0, 15), 'm2',    fast = FALSE),
#                      res_area(seq(0, 15), 'km2',   fast = FALSE)),
#                 left_join,
#                 by = 'res')
# names(h3_res_areas) <-
#   c('h3_resolution', 'avg_area_sqm', 'avg_area_sqkm')
#
# h3_res_els <- dplyr::left_join(res_length(seq(0, 15), 'm', fast = FALSE),
#                                res_length(seq(0, 15), 'km', fast = FALSE), by = 'res')
# names(h3_res_els) <- c('h3_resolution', 'avg_edge_m', 'avg_edge_km')
#
# h3_seps <-
#   dplyr::left_join(res_cendist(seq(0, 15), 'm', fast = FALSE),
#                    res_cendist(seq(0, 15), 'km', fast = FALSE), by = 'res')
# names(h3_seps) <- c('h3_resolution', 'avg_cendist_m', 'avg_cendist_km')
#
# h3_counts <- num_cells(seq(0, 15), fast = FALSE)
# names(h3_counts) <- c('h3_resolution', 'total_unique_indexes')
#
# h3_info_table <-
#   purrr::reduce(list(h3_res_areas, h3_res_els, h3_seps, h3_counts),
#                 left_join, by = 'h3_resolution')
# usethis::use_data(h3_info_table, overwrite = TRUE)
